### PR TITLE
Document supervisor api [ECR-4149]

### PR DESCRIPTION
**Pull request by popzxc**
This PR adds a documentation for public and private REST API for supervisor.

Note that doctests are still sometimes failing due to some bug (which, as occurs, was not really fixed by the recent PR).

### CODE REVIEW

File: services/supervisor/src/api.rs [hunk 1]
```diff
@@ -12,6 +12,578 @@
...
+//!
+//!     - [Request to deploy an artifact](#request-to-deploy-an-artifact)
+//!     - [Request service migration](#request-service-migration)
+//!     - [Request to accept new configuration](#request-to-accept-new-configuration)
+//!     - [Vote for configuration proposal](#vote-for-configuration-proposal)
+//!     - [Obtain current configuration number](#obtain-current-configuration-number)
+//!     - [Obtain current supervisor operating mode](#obtain-current-supervisor-operating-mode)
+//!     - [Check the deployment status](#check-deployment-status)
+//!     - [Check the migration status](#check-migration-status)
+//!
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Nit: Unnecessary newline.

File: services/supervisor/src/api.rs [hunk 2]
```diff
@@ -12,6 +12,578 @@
...
+//! # Public API
+//!
+//! ## Obtaining Consensus Configuration
+//!
+//! | Property    | Value |
+//! |-------------|-------|
+//! | Path        | `/api/supervisor/v1/consensus-config` |
+//! | Method      | GET   |
+//! | Query type  | - |
+//! | Return type | [`BlockInfo`] |
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Shouldn't it be `ConsensusConfig` here?

File: services/supervisor/src/api.rs [hunk 3]
```diff
@@ -12,6 +12,578 @@
...
+//! ```
+//!
+//! ## Obtaining Pending Configuration Proposal
+//!
+//! | Property    | Value |
+//! |-------------|-------|
+//! | Path        | `/api/supervisor/v1/config-proposal` |
+//! | Method      | GET   |
+//! | Query type  | - |
+//! | Return type | `Option<[ConfigProposalWithHash]>` |
```
Comment (slowli) [inline services/supervisor/src/api.rs]: The link here isn't parsed as a link; it probably needs to be written like

```text
`Option<`[`ConfigProposalWithHash`]`>`
```

File: services/supervisor/src/api.rs [hunk 4]
```diff
@@ -12,6 +12,578 @@
...
+//!
+//! Returns the configuration proposal which is currently pending. Returns `None` if there is no
+//! pending configuration at the moment.
+//!
+//! [ConfigProposalWithHash]: ../struct.ConfigProposalWithHash.html
+//!
+//! ```
+//! use exonum_rust_runtime::ServiceFactory;
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Nit: IMO, it would be safe to skip imports / `testkit` initialization if it coincides with the earlier tests:

```rust
let mut testkit = // same as in previous example
let pending_proposal: Option<ConfigProposalWithHash> = testkit
```

File: services/supervisor/src/api.rs [hunk 5]
```diff
@@ -12,6 +12,578 @@
...
+//! Depending on the supervisor operating mode, it may be required to send such a request to
+//! majority of nodes (in "decentralized" mode), or one request will be enough (in "simple" mode).
+//!
+//! After receiving a deployment request, supervisor creates a corresponding transaction, signs it
+//! with node's keys and broadcasts the transaction within the network. The hash of broadcast
+//! transaction is returned from the endpoint.
+//!
+//! For more details on deploy requests, see [crate documentation](../index.html).
+//!
+//! **Warning:** `DeployRequest` structure should be serialized using corresponding protobuf message,
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Off-topic: that `DeployRequest` is serialized into a hex string is unclear from the code sample below. IMO, a better way would be to have a wrapper type like

```rust
pub struct HexString<T>(pub T);

impl<T: BinaryValue> Serialize for HexString<T> { /* ... */ }

// used like
let response = testkit
    .api()
    .private(ApiKind::Service("supervisor"))
    .query(&HexString(deploy_request))
    .post("deploy-artifact")?;
```

This would have the added benefit that `DeployRequest` could have more "sane" serde serialization in other contexts (e.g., if it is returned from the API, or if somebody decides to serialize it in other format, such as `bincode`).

Also, how is a hex string serialized into query params? It could be provided as is for the request body, but as a query it would require the param name. If it *is* serialized as a request body, it makes sense to replace "query type" -> "body type" in the table above. (And maybe update docs in testkit.)

File: services/supervisor/src/api.rs [hunk 6]
```diff
@@ -12,6 +12,578 @@
...
+//! #
+//! # fn main() -> Result<(), failure::Error> {
+//! let mut testkit = TestKitBuilder::validator()
+//!     .with_rust_service(Supervisor)
+//!     .with_artifact(Supervisor.artifact_id())
+//!     .with_instance(Supervisor::simple())
+//!     .with_rust_service(SomeService)
+//!     .create();
+//!
+//! // In this example, we will try to deploy supervisor service.
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Isn't `SomeService` deployed, not a supervisor?

File: services/supervisor/src/api.rs [hunk 7]
```diff
@@ -12,6 +12,578 @@
...
+//! };
+//!
+//! let tx_hash: Hash = testkit
+//!     .api()
+//!     .private(ApiKind::Service("supervisor"))
+//!     .query(&deploy_request)
+//!     .post("deploy-artifact")?;
+//!
+//! let block = testkit.create_block();
+//! let result = block[tx_hash].status();
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Nit: Could this be shortened to

```rust
block[tx_hash].status()?;
```

File: services/supervisor/src/api.rs [hunk 8]
```diff
@@ -12,6 +12,578 @@
...
+//! | Method      | POST   |
+//! | Query type  | [`DeployRequest`] |
+//! | Return type | [`Hash`] |
+//!
+//! Votes for a pending configuration with a certain hash.
+//!
+//! Depending on the supervisor operating mode, it may be required to vote by majority of
+//! nodes (in "decentralized" mode), or one vote will be enough (in "simple" mode).
+//!
+//! The first vote is sent automatically by the node which broadcast the proposal, there is
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Nit: I'd rephrase "The node that broadcast the proposal is considered to have voted for it, there is no need..." because the node does not really *send* the vote, does it?

File: services/supervisor/src/api.rs [hunk 9]
```diff
@@ -258,6 +830,7 @@ impl PublicApi for ApiImpl<'_> {
     }
 }
 
+/// Wires Supervisor API endpoints.
 pub fn wire(builder: &mut ServiceApiBuilder) {
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Does this method need to be public? I'd prefer for it to be `pub(crate)`.

File: services/supervisor/src/lib.rs [hunk 10]
```diff
@@ -157,9 +161,9 @@ use exonum_rust_runtime::{
 
 use crate::{configure::ConfigureMut, mode::Mode};
 
+pub mod api;
```
Comment (slowli) [inline services/supervisor/src/lib.rs]: Nit: IMO, it makes sense not to re-export `api`-related types (i.e., `DeployInfoQuery` and `MigrationInfoQuery`) from the crate root.

File: services/supervisor/src/api.rs (hunk 6 above)
Comment (popzxc) [reply services/supervisor/src/api.rs]: Yep, I forgot to fix the comment.

File: services/supervisor/src/api.rs (hunk 7 above)
Comment (popzxc) [reply services/supervisor/src/api.rs]: I thought that with `assert` the example will be more demonstrative (in other places `?` is used to not-so-important determine places, where we don't want to `unwrap`, but here it's the key point of the example).

File: services/supervisor/src/api.rs (hunk 5 above)
Comment (popzxc) [reply services/supervisor/src/api.rs]: I changed `query` to `body`, and added the notes to the `testkit` docs.

However, I did not introduce `HexString` here, since it's a docs-only PR, and I think it's better to make these changes separately.

File: services/supervisor/src/api.rs (hunk 4 above)
Comment (popzxc) [reply services/supervisor/src/api.rs]: Is it OK now, or should I hide something else as well?

File: services/supervisor/src/api.rs (hunk 5 above)
Comment (slowli) [reply services/supervisor/src/api.rs]: Sure, that's why I prefaced the comment as off-topic.

File: services/supervisor/src/api.rs [hunk 11]
```diff
@@ -79,24 +78,25 @@
 //! | Path        | `/api/supervisor/v1/config-proposal` |
 //! | Method      | GET   |
 //! | Query type  | - |
-//! | Return type | `Option<[ConfigProposalWithHash]>` |
+//! | Return type | Option<[`ConfigProposalWithHash`]> |
```
Comment (slowli) [inline services/supervisor/src/api.rs]: Micro-nit: `Option` is not code-formatted now 🙃
